Remove junk lines caused by redundant bracket #80

nhbinh-uet opened the pull request:
Addressing [ndmitchell/hlint#80](https://github.com/ndmitchell/hlint/issues/80)

Here, we implement a function to format the input and output as suggested from the original issue. As we traverse through the raw hint output strings from both directions to take out all the excess, while keeping both above and below lines for context. This approach has successfully passed all the Hlint test cases and many tests that we had tried. Below are illustrations of how it works when dealing with junk output.

Example with `Redundant bracket`:
```haskell
issue80 = case a of 1 -> (b + 2); 2 -> 100; 3 -> 101; 4 -> 102; 5-> 103
```
Before:
```haskell
Found:
  case a of
    1 -> (b + 2)
    2 -> 100
    3 -> 101
    4 -> 102
    5 -> 103
Perhaps:
  case a of
    1 -> b + 2
    2 -> 100
    3 -> 101
    4 -> 102
    5 -> 103
```
After:
```haskell
Found:
  case a of
    1 -> (b + 2)
    2 -> 100

Perhaps:
  case a of
    1 -> b + 2
    2 -> 100
```
Example with `Monad`:
```haskell
main = do x <- bar; x; y; z; t; f
```
Before:
```haskell
Found:
  do x <- bar
     x
     y
     z
     t
     f
Perhaps:
  do join bar
     y
     z
     t
     f
 ```
 After:
 ```haskell
 Found:
  do x <- bar
     x
     y

Perhaps:
  do join bar
     y
 ```

nhbinh-uet committed:
Update Bracket.hs

epn09 commented on src/Hint/Bracket.hs:
```diff
@@ -190,10 +190,14 @@ bracket pretty isPartialAtom root = f Nothing
         typ = findType v
         r = Replace typ (toSS v) [("x", toSS x)] "x"
+        (from, to) = reduceJunks (pretty o) (Just (pretty (gen x)))
+        reduceJunks :: String -> Maybe String -> (String, Maybe String)
```
Pattern match is not exhaustive. On line 193, you should change `to` to `Just to` so that `reduceJunks` become `String -> String -> (String, String)`.

epn09 commented on src/Hint/Bracket.hs:
```diff
@@ -190,10 +190,14 @@ bracket pretty isPartialAtom root = f Nothing
         typ = findType v
         r = Replace typ (toSS v) [("x", toSS x)] "x"
+        (from, to) = reduceJunks (pretty o) (Just (pretty (gen x)))
+        reduceJunks :: String -> Maybe String -> (String, Maybe String)
```
Pattern match is not exhaustive. On line 193, you should change `to` to `Just to` so that `reduceJunks` become `String -> String -> (String, String)`.

nguyenhaibinh-tpc replied:
> Please give a before-after comparison.

```haskell
main = do
    let a = 1
    let a = 2
    let a = 3
    let a = 4
    let a = (f 7)
    let a = 8
    let a = 9
    let a = 10
    return ()   
  ```
Before:
```
Found:
  do let a = 1    
     let a = 2    
     let a = 3    
     let a = 4    
     let a = 5    
     let a = 6    
     let a = (f 7)
     let a = 8
     let a = 9
     let a = 10
     return ()
Perhaps:
  do let a = 1
     let a = 2
     let a = 3
     let a = 4
     let a = 5
     let a = 6
     let a = f 7
     let a = 8
     let a = 9
     let a = 10
     return ()
```
After:
```
Found:
     let a = (f 7)

Perhaps:
     let a = f 7
 ```

epn09 commented on src/Hint/Bracket.hs:
```diff
@@ -190,10 +190,14 @@ bracket pretty isPartialAtom root = f Nothing
+        (from, to) = reduceJunks (pretty o) (Just (pretty (gen x)))
+        reduceJunks :: String -> Maybe String -> (String, Maybe String)
+        reduceJunks from (Just to) = (unlines st, Just $ unlines nd)
+          where (st, nd) = unzip $ filter (\ (x, y) -> x /= y) $ zip (lines from) (lines to)
```
`\(x, y) -> x /= y` is equivalent to `uncurry (/=)`. (Just a suggestion).

epn09 reviewed (changes requested):
Please give a before-after comparison.

nhbinh-uet committed:
Update Bracket.hs

kk-hainq commented:
Should always mention the original issue: https://github.com/ndmitchell/hlint/issues/80.

Need better PR title, description, and commit message.

Also a lot of testing, either in code, or at least reports in text.

epn09 reviewed (approved):
While this is not what the original issue proposed
> A better approach would be to format both the input and the output as the user wrote it, and just include the line above/line below for context

but it solved the junk problem, so I think it's okay.

nhbinh-uet committed:
Editing a test

Assuming the `do f` in the test is redundant.

kk-hainq commented on src/Hint/Bracket.hs:
```diff
@@ -25,7 +25,7 @@ yes = \ x -> (x && x) -- @Suggestion \x -> x && x
 yes = (`foo` (bar baz)) -- @Suggestion (`foo` bar baz)
 yes = f ((x)) -- @Warning x
-main = do f; (print x) -- @Suggestion do f print x
+main = do f; (print x) -- @Suggestion print x
```
We can't just casually delete `f` out of context. Need better tests, or at least an explanation.

kk-hainq commented on src/Hint/Bracket.hs:
```diff
@@ -190,10 +190,14 @@ bracket pretty isPartialAtom root = f Nothing
         typ = findType v
         r = Replace typ (toSS v) [("x", toSS x)] "x"
+        (from, to) = reduceJunks (pretty o) (pretty (gen x))
+        reduceJunks :: String -> String -> (String, String)
```
We don't need an intermediary function here.
And we have a clean `both` from `import Data.Tuple.Extra` for something like this:
```
(from, to) = both (lines . pretty) (o, gen x)
(from', to') = both unlines $ unzip $ filter (uncurry (/=)) (zip from to)
```

kk-hainq reviewed (changes requested):
See comments. We need better tests and documentation in general. 

Also, does this happen in other cases as well? Or just with `Bracket`?

nhbinh-uet committed:
#80 Remove redundant code in hints output

Added a function to clear the junk input/output, only left a line above and below for context.

nhbinh-uet committed:
Revert "#80 Remove redundant code in hints output"

This reverts commit a90c041361b24d16b8a3e93acdd15d4e9c928ead.

nhbinh-uet committed:
Added the missing packages

Added the that packages were removed by mistake

nhbinh-uet committed:
Revert "Revert "#80 Remove redundant code in hints output""

This reverts commit 85553f93f7023e74c810fa3307b39159ea2b2d65.

kk-hainq commented on src/Hint/Bracket.hs:
```diff
@@ -266,4 +266,4 @@ dollar = concatMap f . universe
 splitInfix (L l (OpApp _ lhs op rhs)) =
   [(L l . OpApp noExtField lhs op, rhs), (\lhs -> L l (OpApp noExtField lhs op rhs), lhs)]
-splitInfix _ = []
+splitInfix _ = []
```
Remove this change. And remember to add EOF(s).

nhbinh-uet committed:
Added EOF to Bracket.hs

nhbinh-uet committed:
Added EOF to Idea.hs

epn09 commented on src/Idea.hs:
```diff
@@ -74,14 +76,25 @@ showIdeaANSI = showEx hsColourConsole
+        go (_, []) = 0
+        go (f : fs, t : ts) = if f == t then 1 + go (fs, ts) else 0
+        (start, end) = ( 1 `max` go tup - 1, 1 `max` go rev_tup)
+    in both (unlines.(\xs-> (xs !!) <$> [start..(length xs - end)])) tup
```
This `reduceJunk` is too abstract and it's really difficult to figure out what it does just by looking at the code.

`(xs !!) <$> [a..b]` is also not a good way to extract a sub-array from index `a` to `b` because its complexity is O(b^2).

epn09 commented on src/Idea.hs:
```diff
@@ -18,6 +18,8 @@ import Prelude
 import GHC.Types.SrcLoc
 import GHC.Utils.Outputable
 import GHC.Util
+import Data.Maybe (fromMaybe, isNothing, fromJust)
```
These imports aren't used.

epn09 reviewed (changes requested):
Please make `reducedJunk` easier to understand.

nhbinh-uet committed:
Improve and explicate Idea.hs

- Explanations added
- reduceJunk function reconstructed
- Improve the complexity of the subarray extracting function
- Remove redundant imports

epn09 commented on src/Idea.hs:
```diff
@@ -79,22 +78,28 @@ showEx tt Idea{..} = unlines $
+reduceJunk :: (String, String) -> (String, String)
+reduceJunk x@(from, to) =
+    let 
+        tup = both lines x; rev_tup = both (reverse. lines) x
```
You should reuse `tup` in the definition of `rev_tup`.

epn09 commented on src/Idea.hs:
```diff
@@ -79,22 +78,28 @@ showEx tt Idea{..} = unlines $
+reduceJunk x@(from, to) =
+    let 
+        tup = both lines x; rev_tup = both (reverse. lines) x
+        -- compare each line to get the first dissimilar position
```
This `go` function returns the length of the longest common prefix? If so, please change its name to something that reminds reader of LCP.

epn09 commented on src/Idea.hs:
```diff
@@ -79,22 +78,28 @@ showEx tt Idea{..} = unlines $
+        -- subtract 1 in both version to get the corresponding above/below line for context.
+        (start, end) = (go tup - 1, go rev_tup - 1)      
+        -- extract a sub-array from l to r where r is the index in reversed version
+        extract :: [String] -> Int -> Int -> [String]
```
A signature like `[a] -> Int -> Int -> [a]`  might be better. In addition, because `drop` and `take` both have the type of `Int -> [a] -> [a]`, you might want to reorder arguments of this function to make its type `Int -> Int -> [a] -> [a]`.

epn09 commented on src/Idea.hs:
```diff
@@ -79,22 +78,28 @@ showEx tt Idea{..} = unlines $
+        (start, end) = (go tup - 1, go rev_tup - 1)      
+        -- extract a sub-array from l to r where r is the index in reversed version
+        extract :: [String] -> Int -> Int -> [String]
+        extract hints l r = drop l $ take (length hints - r) hints
```
It's possible to avoid using `length hints - r`, please try using `l` and `r` only. Then, you can also make this function point-free.

nhbinh-uet replied:
```haskell
extract = (. drop) . (.) . dropSuffix
```
I can change the function to this with an extra `dropSuffix` function, but I'm not sure whether it's more intelligible

epn09 replied:
How's about this?
```haskell
extract l r = drop l . take r
```

nhbinh-uet replied:
Sorry for the unclear code, but the extract function here is actually removing the first `l` elements and the last `r` elements. Should I change it to get rid of the confusion?

epn09 commented on src/Idea.hs:
```diff
@@ -79,22 +78,28 @@ showEx tt Idea{..} = unlines $
     ["Note: " ++ n | let n = showNotes ideaNote, n /= ""]
     where
-        (ideaFrom', ideaTo') = maybe (ideaFrom, Nothing) (second Just.(reduceJunk ideaFrom)) ideaTo
+        (ideaFrom', ideaTo') = maybe (ideaFrom, Nothing) ((second Just) . reduceJunk . (ideaFrom, )) ideaTo
```
Why did you make this change? I think the original one is good enough.

nhbinh-uet committed:
Update Idea.hs

Changes to functions name and unclear variables.

epn09 reviewed (commented):
I think now the code is clear enough. What's your opinion, @kk-hainq?
